File: backend/trips/models/trip.py
import uuid

from django.db import models


class Trip(models.Model):
    id = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
    driver = models.ForeignKey(
        "trips.Driver", on_delete=models.SET_NULL, null=True, blank=True
    )
    vehicle = models.ForeignKey(
        "trips.Vehicle", on_delete=models.SET_NULL, null=True, blank=True
    )

    current_location_label = models.CharField(max_length=300)
    current_location_lat = models.FloatField()
    current_location_lng = models.FloatField()

    pickup_label = models.CharField(max_length=300)
    pickup_lat = models.FloatField()
    pickup_lng = models.FloatField()

    dropoff_label = models.CharField(max_length=300)
    dropoff_lat = models.FloatField()
    dropoff_lng = models.FloatField()

    current_cycle_used_hours = models.FloatField(
        help_text="Hours already used in the rolling 8-day cycle (0-70)"
    )
    start_datetime = models.DateTimeField()

    total_distance_miles = models.FloatField(null=True, blank=True)
    total_driving_minutes = models.IntegerField(null=True, blank=True)
    total_trip_minutes = models.IntegerField(null=True, blank=True)
    route_geometry = models.JSONField(
        null=True, blank=True, help_text="GeoJSON LineString or encoded polyline"
    )

    shipper_name = models.CharField(max_length=200, blank=True, default="")
    commodity = models.CharField(max_length=200, blank=True, default="")
    bol_number = models.CharField(max_length=100, blank=True, default="")

    created_at = models.DateTimeField(auto_now_add=True)

    class Meta:
        ordering = ["-created_at"]

    def __str__(self):
        return f"Trip {self.id} ({self.current_location_label} → {self.dropoff_label})"

File: frontend/src/types/trip.ts
export interface Driver {
  id: number;
  name: string;
  license_number: string;
  carrier_name: string;
  carrier_main_office: string;
  home_terminal_tz: string;
}

export interface Vehicle {
  id: number;
  truck_number: string;
  trailer_number: string;
  license_plate: string;
  license_state: string;
}

export type StopType =
  | "origin"
  | "pickup"
  | "dropoff"
  | "fuel"
  | "break_30min"
  | "rest_10hr"
  | "restart_34hr"
  | "post_trip";

export type DutyStatus =
  | "off_duty"
  | "sleeper_berth"
  | "driving"
  | "on_duty_not_driving";

export interface RouteStop {
  id: number;
  sequence: number;
  stop_type: StopType;
  label: string;
  lat: number;
  lng: number;
  arrival_time: string;
  departure_time: string;
  duration_minutes: number;
  cumulative_miles: number;
  notes: string;
}

export interface DutyStatusEvent {
  id: number;
  start_time: string;
  end_time: string;
  status: DutyStatus;
  location_label: string;
  lat: number;
  lng: number;
  remark: string;
}

export interface DailyLog {
  id: number;
  log_date: string;
  total_miles_driving_today: number;
  total_off_duty: number;
  total_sleeper_berth: number;
  total_driving: number;
  total_on_duty_not_driving: number;
  shipping_doc_number: string;
  svg_url: string;
}

export interface Trip {
  id: string;
  current_location_label: string;
  current_location_lat: number;
  current_location_lng: number;
  pickup_label: string;
  pickup_lat: number;
  pickup_lng: number;
  dropoff_label: string;
  dropoff_lat: number;
  dropoff_lng: number;
  current_cycle_used_hours: number;
  start_datetime: string;
  total_distance_miles: number | null;
  total_driving_minutes: number | null;
  total_trip_minutes: number | null;
  route_geometry: object | null;
  shipper_name: string;
  commodity: string;
  bol_number: string;
  created_at: string;
  stops: RouteStop[];
  duty_events: DutyStatusEvent[];
  daily_logs: DailyLog[];
}

export interface TripCreatePayload {
  current_location: string;
  pickup_location: string;
  dropoff_location: string;
  current_cycle_used_hours: number;
  start_datetime?: string;
  driver_name?: string;
  carrier_name?: string;
  main_office?: string;
  truck_number?: string;
  trailer_number?: string;
  co_driver?: string;
  shipper_name?: string;
  commodity?: string;
  bol_number?: string;
}
